src/assign.rs

```diff
@@ -58,25 +54,456 @@ pub struct Assignment<'a> {
+    FailedToParseIndex {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`ParseIndexError`]
+        source: ParseIndexError,
+    },
+    /// An `Index` `Token` within the `Pointer` was out of bounds of the
+    /// target array.
+    OutOfBounds {
+        /// Offset of the partial pointer starting with the invalid index.
+        offset: usize,
+        /// The source [`OutOfBoundsError`]
+        source: OutOfBoundsError,
+    },
+}
+impl fmt::Display for AssignError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::FailedToParseIndex { offset, .. } => {
+                write!(f, "failed to parse index at offset {}", offset)
+            }
+            Self::OutOfBounds { offset, .. } => {
+                write!(f, "index at offset {} out of bounds", offset)
+            }
+        }
+    }
+}
+#[cfg(feature = "std")]
+impl std::error::Error for AssignError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::FailedToParseIndex { source, .. } => Some(source),
+            Self::OutOfBounds { source, .. } => Some(source),
+        }
+    }
+}
+
+enum Assigned<'v> {
+    Done(Assignment<'v>),
+    Continue {
+        next_buf: PointerBuf,
+        next_value: &'v mut Value,
+        same_src: Value,
+    },
+}
+
+fn assign_array<'v>(
+    token: Token<'_>,
+    remaining: &Pointer,
+    mut buf: PointerBuf,
+    array: &'v mut Vec<Value>,
+    src: Value,
+    offset: usize,
+) -> Result<Assigned<'v>, AssignError> {
+    // parsing the index
+    let idx = token
+        .to_index()
+        .map_err(|source| AssignError::FailedToParseIndex { offset, source })?
+        .for_len_incl(array.len())
+        .map_err(|source| AssignError::OutOfBounds { offset, source })?;
+    buf.push_back(idx.into());
+
+    debug_assert!(idx <= array.len());
+
+    if idx < array.len() {
+        // element exists in the array, we either need to replace it or continue
+        // depending on whether this is the last token or not
+        if remaining.is_root() {
+            // last token, we replace the value and call it a day
+            let replaced = Some(mem::replace(&mut array[idx], src));
+            let assigned = &mut array[idx];
+            Ok(Assigned::Done(Assignment {
+                assigned,
+                assigned_to: buf,
+                replaced,
+            }))
+        } else {
+            // not the last token, we continue with a mut ref to the element as
+            // the next value
+            Ok(Assigned::Continue {
+                next_value: &mut array[idx],
+                same_src: src,
+                next_buf: buf,
+            })
+        }
+    } else {
+        // element does not exist in the array.
+        // we create the path and assign the value
+        let src = expand_src_path(remaining, src)?;
```

Comment:
Hmm I think there's an issue here if the token contains a non-zero number - it'll create a map in that case.

Reply:
I think that's okay, sort of. Any non `"0"` index is going to be turned into an object too. 
e.g., this is valid:
```json
{ "-": "valid json", "0": "still valid" }

I've been kind of fretting over it because an intended object with keys `"-"` or `"0"` is going to spit out an array. So if the keys are dynamic and someone uses one of those as keys and assumptions are made 😭 . There just really isn't any clean way around it though and its the only way to properly allow for auto-expansion.

Perhaps that's the answer. I'm not sure.

Reply:
Hmm yeah, we've faced the same problem when using `jsonptr` to access a different JSON-like data structure. A possible solution here could be to allow the caller to choose a strategy, and a good default is probably to just not expand at all (other than when it's an unambiguous append-to-array situation).

Reply:
I considered a strategy, but you'd either need to opt in to all objects or all possible arrays. Only the first option would prevent accidentally getting the wrong shape of data. I guess the `Strategy` could be something to the effect of
```rust
pub enum Expansion {
    Auto,
    Error
}
```